File: vane-core/src/main/java/org/oddlama/vane/util/LazyLocation.java
package org.oddlama.vane.util;

import java.util.UUID;
import org.bukkit.Bukkit;
import org.bukkit.Location;

public class LazyLocation {

    private final UUID worldId;
    private Location location;

    public LazyLocation(final Location location) {
        this.worldId = location.getWorld() == null ? null : location.getWorld().getUID();
        this.location = location.clone();
    }

    public LazyLocation(final UUID worldId, double x, double y, double z, float pitch, float yaw) {
        this.worldId = worldId;
        this.location = new Location(null, x, y, z, pitch, yaw);
    }

    public UUID worldId() {
        return worldId;
    }

    public Location location() {
        if (worldId != null && location.getWorld() == null) {
            location.setWorld(Bukkit.getWorld(worldId));
        }

        return location;
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/util/LazyLocation.kt
package org.oddlama.vane.util

import org.bukkit.Bukkit
import org.bukkit.Location
import java.util.*

/**
 * Lazily resolves world references for stored locations.
 *
 * @param worldId world UUID used for lazy world lookup.
 * @param location backing location instance.
 */
class LazyLocation(
    /** World UUID used for lazy world resolution. */
    val worldId: UUID?,
    /** Backing location value. */
    private val location: Location,
) {
    /**
     * Creates a lazy location from an existing location.
     */
    constructor(location: Location) : this(
        worldId = location.world?.uid,
        location = location.clone(),
    )

    /**
     * Creates a lazy location from explicit coordinates and world UUID.
     */
    constructor(worldId: UUID?, x: Double, y: Double, z: Double, pitch: Float, yaw: Float) : this(
        worldId = worldId,
        location = Location(null, x, y, z, pitch, yaw),
    )

    /**
     * Returns the resolved location, loading the world lazily when needed.
     */
    fun location(): Location {
        if (worldId != null && location.world == null) {
            location.world = Bukkit.getWorld(worldId)
        }
        return location
    }
}
